### PR TITLE
perlPackages: cross-compilation fixes

### LINKED ISSUE #36675: perlPackages.XMLParser fails to cross-compile
## Issue description

Cross-compiling `perlPackages.XMLParser` on nixpkgs' `staging` branch (e.g. with @shlevy's perl cross-compilation patch, 306d5cdf03ad6375861a20b8885ef38699ca3c23) fails.

### Steps to reproduce

```
$ git -C nixpkgs show
commit 306d5cdf03ad6375861a20b8885ef38699ca3c23
...
$ cat test.nix
let
  nixpkgs = import ./nixpkgs {};
  crossPkgs = import ./nixpkgs { crossSystem = nixpkgs.lib.systems.examples.armv7l-hf-multiplatform; };
in crossPkgs.perlPackages.XMLParser
$ nix-build ./test.nix
...
patching sources
configuring
patching ./samples/xmlfilter...
patching ./samples/canonical...
patching ./samples/xmlstats...
patching ./samples/xmlcomments...

Expat must be installed prior to building XML::Parser and I can't find
it in the standard library directories. Install 'expat-devel' package with your
OS package manager. See 'README'.

Or you can download expat from:

http://sourceforge.net/projects/expat/

If expat is installed, but in a non-standard directory, then use the
following options to Makefile.PL:

    EXPATLIBPATH=...  To set the directory in which to find libexpat

    EXPATINCPATH=...  To set the directory in which to find expat.h

For example:

    perl Makefile.PL EXPATLIBPATH=/home/me/lib EXPATINCPATH=/home/me/include

Note that if you build against a shareable library in a non-standard location
you may (on some platforms) also have to set your LD_LIBRARY_PATH environment
variable at run time for perl to find the library.

Checking if your kit is complete...
Warning: prerequisite LWP 0 not found.
Looks good
ERROR from evaluation of /tmp/nix-build-perl-XML-Parser-2.41-arm-unknown-linux-gnueabihf.drv-0/XML-Parser-2.41/Expat/Makefile.PL: Can't load module Tie::Hash::NamedCapture, dynamic loading not available in this perl.
  (You may need to build a new perl executable which either supports
  dynamic loading or has the Tie::Hash::NamedCapture module statically linked into it.)
 at /nix/store/1v96mbpfg5g9n58b74q5yafdfbhl1kj9-perl-5.24.3-arm-unknown-linux-gnueabihf/lib/perl5/5.24.3/English.pm line 155.
Compilation failed in require at /nix/store/1v96mbpfg5g9n58b74q5yafdfbhl1kj9-perl-5.24.3-arm-unknown-linux-gnueabihf/lib/perl5/5.24.3/English.pm line 155.
BEGIN failed--compilation aborted at /nix/store/1v96mbpfg5g9n58b74q5yafdfbhl1kj9-perl-5.24.3-arm-unknown-linux-gnueabihf/lib/perl5/5.24.3/English.pm line 155.
Compilation failed in require at ./Makefile.PL line 3.
BEGIN failed--compilation aborted at ./Makefile.PL line 3.
builder for '/nix/store/g4qzia34nq5r4zizhp3vhl712lavqrkd-perl-XML-Parser-2.41-arm-unknown-linux-gnueabihf.drv' failed with exit code 255
error: build of '/nix/store/g4qzia34nq5r4zizhp3vhl712lavqrkd-perl-XML-Parser-2.41-arm-unknown-linux-gnueabihf.drv' failed
```

## Technical details

Hmmm...
```
$ nix-shell -p nix-info --run "nix-info -m"
error: undefined variable 'nix-info' at (string):1:94
(use '--show-trace' to show detailed location information)
```

### PR DESCRIPTION
**Pull request by volth**
###### Motivation for this change

https://github.com/NixOS/nixpkgs/pull/48241 + 
https://github.com/NixOS/nixpkgs/pull/48242 +
https://github.com/NixOS/nixpkgs/pull/48246 rebased to staging 

Needed for cross-compilation of whole closures (as dependency of switch-to-configuration.pl)

Fixes #36675

### CODE REVIEW

**Comment by volth** [comment]
@GrahamcOfBorg build pkgsCross.armv7l-hf-multiplatform.perlPackages.XMLParser

File: pkgs/top-level/perl-packages.nix [hunk 1]
```diff
@@ -17596,9 +17600,11 @@ let
       url = mirror://cpan/authors/id/T/TO/TODDR/XML-Parser-2.44.tar.gz;
       sha256 = "05ij0g6bfn27iaggxf8nl5rhlwx6f6p6xmdav6rjcly3x5zd1s8s";
     };
-    patchPhase = if stdenv.isCygwin then ''
+    patchPhase = stdenv.lib.optionalString (stdenv.buildPlatform != stdenv.hostPlatform) ''
+      substituteInPlace Expat/Makefile.PL --replace 'use English;' '#'
```
Comment (Mic92) [inline pkgs/top-level/perl-packages.nix]: Is this module not used at all in the file? I am surprised that it works without it.

Comment (volth) [reply pkgs/top-level/perl-packages.nix]: It is Makefile, so it is used only at build time; if it did not crash during the build then it works.

**Comment by volth** [comment]
@GrahamcOfBorg build pkgsCross.armv7l-hf-multiplatform.perlPackages.LWP

**Pull request merged**